File: python2/sagerupdate.py
import numpy as np
def sagerupdate(image_vec, weights, lr):
    '''

    
    %SAGERUPDATE updates the weights of the connections between input pixels
%and output units given the presentation of an image according to Sanger's
%rule.
    '''
    num_pixels = len(image_vec)
    weight_size = weights.shape
    num_units = weight_size[0]
    output_values = np.dot(weights ,image_vec)
#     for oja_input in image_vec:
#         for j in range
#         pass
    if 0:
        '''
            [TBC] performance bottle-neck
        '''
#         assert 0,[(x,eval(x).shape)  for x in ("image_vec",
#                                                "output_values",
#                                                "weights")]
        for i in range(num_pixels):
            oja_input = image_vec[i]
            for j in range(num_units):
                oja_input = oja_input - output_values[j]*weights[j,i]
                weights[j,i] = weights[j,i] + lr * output_values[j] * oja_input
    else:
        '''vectorised approximation'''
        
        oja_input = image_vec
        recov = np.cumsum(output_values * weights,axis=0)
        weights += lr * output_values * (oja_input.T - recov)
#     for j in range(num_units):
#         output_values * weights
    return weights

File: matlab/sagerupdate.m
function [ weights ] = sagerupdate( image_vec, weights, learn_rate )
%SAGERUPDATE updates the weights of the connections between input pixels
%and output units given the presentation of an image according to Sanger's
%rule.

%Extract number of pixels in image.
num_pixels=length(image_vec);

%Extract number of output units.
weights_size=size(weights);
num_units=weights_size(1);

%Calculate output values.
output_values=weights*image_vec;

%Update weights using algorithm suggested by Sager
for input=1:num_pixels
    oja_input=image_vec(input);
    for J=1:num_units
        oja_input=oja_input-output_values(J)*weights(J,input);
        weights(J,input)=weights(J,input)+learn_rate*output_values(J)*oja_input;
    end
end   


end

